File: library/src/main/java/org/xmtp/android/library/Contacts.kt
package org.xmtp.android.library

import kotlinx.coroutines.runBlocking
import org.xmtp.android.library.messages.ContactBundle
import org.xmtp.android.library.messages.ContactBundleBuilder
import org.xmtp.android.library.messages.EnvelopeBuilder
import org.xmtp.android.library.messages.Pagination
import org.xmtp.android.library.messages.Topic
import org.xmtp.android.library.messages.walletAddress
import org.xmtp.proto.message.api.v1.MessageApiOuterClass
import org.xmtp.proto.message.contents.PrivatePreferences.PrivatePreferencesAction
import uniffi.xmtpv3.FfiConsent
import uniffi.xmtpv3.FfiConsentEntityType
import uniffi.xmtpv3.FfiConsentState
import java.util.Date

enum class ConsentState {
    ALLOWED,
    DENIED,
    UNKNOWN;

    companion object {
        fun toFfiConsentState(option: ConsentState): FfiConsentState {
            return when (option) {
                ConsentState.ALLOWED -> FfiConsentState.ALLOWED
                ConsentState.DENIED -> FfiConsentState.DENIED
                else -> FfiConsentState.UNKNOWN
            }
        }

        fun fromFfiConsentState(option: FfiConsentState): ConsentState {
            return when (option) {
                FfiConsentState.ALLOWED -> ConsentState.ALLOWED
                FfiConsentState.DENIED -> ConsentState.DENIED
                else -> ConsentState.UNKNOWN
            }
        }
    }
}

enum class EntryType {
    ADDRESS,
    GROUP_ID,
    INBOX_ID;

    companion object {
        fun toFfiConsentEntityType(option: EntryType): FfiConsentEntityType {
            return when (option) {
                EntryType.ADDRESS -> FfiConsentEntityType.ADDRESS
                EntryType.GROUP_ID -> FfiConsentEntityType.CONVERSATION_ID
                EntryType.INBOX_ID -> FfiConsentEntityType.INBOX_ID
            }
        }

        fun fromFfiConsentEntityType(option: FfiConsentEntityType): EntryType {
            return when (option) {
                FfiConsentEntityType.ADDRESS -> EntryType.ADDRESS
                FfiConsentEntityType.CONVERSATION_ID -> EntryType.GROUP_ID
                FfiConsentEntityType.INBOX_ID -> EntryType.INBOX_ID
            }
        }
    }
}

data class ConsentListEntry(
    val value: String,
    val entryType: EntryType,
    val consentType: ConsentState,
) {
    companion object {
        fun address(
            address: String,
            type: ConsentState = ConsentState.UNKNOWN,
        ): ConsentListEntry {
            return ConsentListEntry(address, EntryType.ADDRESS, type)
        }

        fun groupId(
            groupId: String,
            type: ConsentState = ConsentState.UNKNOWN,
        ): ConsentListEntry {
            return ConsentListEntry(groupId, EntryType.GROUP_ID, type)
        }

        fun inboxId(
            inboxId: String,
            type: ConsentState = ConsentState.UNKNOWN,
        ): ConsentListEntry {
            return ConsentListEntry(inboxId, EntryType.INBOX_ID, type)
        }
    }

    fun toFfiConsent(): FfiConsent {
        return FfiConsent(
            EntryType.toFfiConsentEntityType(entryType),
            ConsentState.toFfiConsentState(consentType),
            value
        )
    }

    val key: String
        get() = "${entryType.name}-$value"
}

class ConsentList(
    val client: Client,
    val entries: MutableMap<String, ConsentListEntry> = mutableMapOf(),
) {
    private var lastFetched: Date? = null

    @OptIn(ExperimentalUnsignedTypes::class)
    suspend fun load(): List<ConsentListEntry> {
        if (client.hasV2Client) {
            val newDate = Date()
            val publicKey =
                client.v1keys.identityKey.publicKey.secp256K1Uncompressed.bytes
            val privateKey = client.v1keys.identityKey.secp256K1.bytes
            val identifier: String =
                uniffi.xmtpv3.generatePrivatePreferencesTopicIdentifier(
                    privateKey.toByteArray(),
                )
            val envelopes =
                client.apiClient!!.envelopes(
                    Topic.preferenceList(identifier).description,
                    Pagination(
                        after = lastFetched,
                        direction = MessageApiOuterClass.SortDirection.SORT_DIRECTION_ASCENDING,
                        limit = 500
                    ),
                )

            lastFetched = newDate
            val preferences: MutableList<PrivatePreferencesAction> = mutableListOf()
            for (envelope in envelopes) {
                val payload =
                    uniffi.xmtpv3.userPreferencesDecrypt(
                        publicKey.toByteArray(),
                        privateKey.toByteArray(),
                        envelope.message.toByteArray(),
                    )

                preferences.add(
                    PrivatePreferencesAction.parseFrom(
                        payload.toUByteArray().toByteArray(),
                    )
                )
            }

            preferences.iterator().forEach { preference ->
                preference.allowAddress?.walletAddressesList?.forEach { address ->
                    allow(address)
                }
                preference.denyAddress?.walletAddressesList?.forEach { address ->
                    deny(address)
                }
                preference.allowGroup?.groupIdsList?.forEach { groupId ->
                    allowGroup(groupId)
                }
                preference.denyGroup?.groupIdsList?.forEach { groupId ->
                    denyGroup(groupId)
                }

                preference.allowInboxId?.inboxIdsList?.forEach { inboxId ->
                    allowInboxId(inboxId)
                }
                preference.denyInboxId?.inboxIdsList?.forEach { inboxId ->
                    denyInboxId(inboxId)
                }
            }
        }
        return entries.values.toList()
    }

    suspend fun publish(entries: List<ConsentListEntry>) {
        if (client.v3Client != null) {
            setV3ConsentState(entries)
        }
        if (client.hasV2Client) {
            val payload = PrivatePreferencesAction.newBuilder().also {
                entries.iterator().forEach { entry ->
                    when (entry.entryType to entry.consentType) {
                        EntryType.ADDRESS to ConsentState.ALLOWED -> it.setAllowAddress(
                            PrivatePreferencesAction.AllowAddress.newBuilder()
                                .addWalletAddresses(entry.value)
                        )

                        EntryType.ADDRESS to ConsentState.DENIED -> it.setDenyAddress(
                            PrivatePreferencesAction.DenyAddress.newBuilder()
                                .addWalletAddresses(entry.value)
                        )

                        EntryType.GROUP_ID to ConsentState.ALLOWED -> it.setAllowGroup(
                            PrivatePreferencesAction.AllowGroup.newBuilder()
                                .addGroupIds(entry.value)
                        )

                        EntryType.GROUP_ID to ConsentState.DENIED -> it.setDenyGroup(
                            PrivatePreferencesAction.DenyGroup.newBuilder().addGroupIds(entry.value)
                        )

                        EntryType.INBOX_ID to ConsentState.ALLOWED -> it.setAllowInboxId(
                            PrivatePreferencesAction.AllowInboxId.newBuilder()
                                .addInboxIds(entry.value)
                        )

                        EntryType.INBOX_ID to ConsentState.DENIED -> it.setDenyInboxId(
                            PrivatePreferencesAction.DenyInboxId.newBuilder()
                                .addInboxIds(entry.value)
                        )

                        else -> it.clearMessageType()
                    }
                }
            }.build()

            val publicKey =
                client.v1keys.identityKey.publicKey.secp256K1Uncompressed.bytes
            val privateKey = client.v1keys.identityKey.secp256K1.bytes
            val identifier: String =
                uniffi.xmtpv3.generatePrivatePreferencesTopicIdentifier(
                    privateKey.toByteArray(),
                )

            val message =
                uniffi.xmtpv3.userPreferencesEncrypt(
                    publicKey.toByteArray(),
                    privateKey.toByteArray(),
                    payload.toByteArray(),
                )

            val envelope = EnvelopeBuilder.buildFromTopic(
                Topic.preferenceList(identifier),
                Date(),
                ByteArray(message.size) { message[it] },
            )

            client.publish(listOf(envelope))
        }
    }

    suspend fun setV3ConsentState(entries: List<ConsentListEntry>) {
        client.v3Client?.setConsentStates(entries.map { it.toFfiConsent() })
    }

    fun allow(address: String): ConsentListEntry {
        val entry = ConsentListEntry.address(address, ConsentState.ALLOWED)
        entries[entry.key] = entry

        return entry
    }

    fun deny(address: String): ConsentListEntry {
        val entry = ConsentListEntry.address(address, ConsentState.DENIED)
        entries[entry.key] = entry

        return entry
    }

    fun allowGroup(groupId: String): ConsentListEntry {
        val entry = ConsentListEntry.groupId(groupId, ConsentState.ALLOWED)
        entries[entry.key] = entry

        return entry
    }

    fun denyGroup(groupId: String): ConsentListEntry {
        val entry = ConsentListEntry.groupId(groupId, ConsentState.DENIED)
        entries[entry.key] = entry

        return entry
    }

    fun allowInboxId(inboxId: String): ConsentListEntry {
        val entry = ConsentListEntry.inboxId(inboxId, ConsentState.ALLOWED)
        entries[entry.key] = entry

        return entry
    }

    fun denyInboxId(inboxId: String): ConsentListEntry {
        val entry = ConsentListEntry.inboxId(inboxId, ConsentState.DENIED)
        entries[entry.key] = entry

        return entry
    }

    suspend fun state(address: String): ConsentState {
        client.v3Client?.let {
            return ConsentState.fromFfiConsentState(
                it.getConsentState(
                    FfiConsentEntityType.ADDRESS,
                    address
                )
            )
        }
        val entry = entries[ConsentListEntry.address(address).key]
        return entry?.consentType ?: ConsentState.UNKNOWN
    }

    suspend fun groupState(groupId: String): ConsentState {
        client.v3Client?.let {
            return ConsentState.fromFfiConsentState(
                it.getConsentState(
                    FfiConsentEntityType.CONVERSATION_ID,
                    groupId
                )
            )
        }
        val entry = entries[ConsentListEntry.groupId(groupId).key]
        return entry?.consentType ?: ConsentState.UNKNOWN
    }

    suspend fun inboxIdState(inboxId: String): ConsentState {
        client.v3Client?.let {
            return ConsentState.fromFfiConsentState(
                it.getConsentState(
                    FfiConsentEntityType.INBOX_ID,
                    inboxId
                )
            )
        }
        val entry = entries[ConsentListEntry.inboxId(inboxId).key]
        return entry?.consentType ?: ConsentState.UNKNOWN
    }
}

data class Contacts(
    var client: Client,
    val knownBundles: MutableMap<String, ContactBundle> = mutableMapOf(),
    val hasIntroduced: MutableMap<String, Boolean> = mutableMapOf(),
    var consentList: ConsentList = ConsentList(client),
) {

    suspend fun refreshConsentList(): ConsentList {
        val entries = consentList.load()
        consentList.setV3ConsentState(entries)
        return consentList
    }

    suspend fun allow(addresses: List<String>) {
        val entries = addresses.map {
            consentList.allow(it)
        }
        consentList.publish(entries)
    }

    suspend fun deny(addresses: List<String>) {
        val entries = addresses.map {
            consentList.deny(it)
        }
        consentList.publish(entries)
    }

    suspend fun allowGroups(groupIds: List<String>) {
        val entries = groupIds.map {
            consentList.allowGroup(it)
        }
        consentList.publish(entries)
    }

    suspend fun denyGroups(groupIds: List<String>) {
        val entries = groupIds.map {
            consentList.denyGroup(it)
        }
        consentList.publish(entries)
    }

    suspend fun allowInboxes(inboxIds: List<String>) {
        val entries = inboxIds.map {
            consentList.allowInboxId(it)
        }
        consentList.publish(entries)
    }

    suspend fun denyInboxes(inboxIds: List<String>) {
        val entries = inboxIds.map {
            consentList.denyInboxId(it)
        }
        consentList.publish(entries)
    }

    suspend fun isAllowed(address: String): Boolean {
        return consentList.state(address) == ConsentState.ALLOWED
    }

    suspend fun isDenied(address: String): Boolean {
        return consentList.state(address) == ConsentState.DENIED
    }

    suspend fun isGroupAllowed(groupId: String): Boolean {
        return consentList.groupState(groupId) == ConsentState.ALLOWED
    }

    suspend fun isGroupDenied(groupId: String): Boolean {
        return consentList.groupState(groupId) == ConsentState.DENIED
    }

    suspend fun isInboxAllowed(inboxId: String): Boolean {
        return consentList.inboxIdState(inboxId) == ConsentState.ALLOWED
    }

    suspend fun isInboxDenied(inboxId: String): Boolean {
        return consentList.inboxIdState(inboxId) == ConsentState.DENIED
    }

    fun has(peerAddress: String): Boolean = knownBundles[peerAddress] != null

    fun needsIntroduction(peerAddress: String): Boolean = hasIntroduced[peerAddress] != true

    fun find(peerAddress: String): ContactBundle? {
        val knownBundle = knownBundles[peerAddress]
        if (knownBundle != null) {
            return knownBundle
        }
        val response = runBlocking { client.query(topic = Topic.contact(peerAddress)) }

        if (response.envelopesList.isNullOrEmpty()) return null

        for (envelope in response.envelopesList) {
            val contactBundle = ContactBundleBuilder.buildFromEnvelope(envelope)
            knownBundles[peerAddress] = contactBundle
            val address = contactBundle.walletAddress
            if (address?.lowercase() == peerAddress.lowercase()) {
                return contactBundle
            }
        }

        return null
    }
}

File: Sources/XMTPiOS/Contacts.swift
//
//  Contacts.swift
//
//
//  Created by Pat Nakajima on 12/8/22.
//

import Foundation
import LibXMTP

public typealias PrivatePreferencesAction = Xmtp_MessageContents_PrivatePreferencesAction

public enum ConsentState: String, Codable {
	case allowed, denied, unknown
}

public struct ConsentListEntry: Codable, Hashable {
	public enum EntryType: String, Codable {
		case address, group_id, inbox_id
	}

	static func address(_ address: String, type: ConsentState = .unknown) -> ConsentListEntry {
		ConsentListEntry(value: address, entryType: .address, consentType: type)
	}
	
	static func groupId(groupId: String, type: ConsentState = ConsentState.unknown) -> ConsentListEntry {
		ConsentListEntry(value: groupId, entryType: .group_id, consentType: type)
	}
	
	static func inboxId(_ inboxId: String, type: ConsentState = .unknown) -> ConsentListEntry {
		ConsentListEntry(value: inboxId, entryType: .inbox_id, consentType: type)
	}

	public var value: String
	public var entryType: EntryType
	public var consentType: ConsentState

	var key: String {
		"\(entryType)-\(value)"
	}
}

public enum ContactError: Error {
	case invalidIdentifier
}

public actor EntriesManager {
	public var map: [String: ConsentListEntry] = [:]

	func set(_ key: String, _ object: ConsentListEntry) {
		map[key] = object
	}

	func get(_ key: String) -> ConsentListEntry? {
		map[key]
	}
}

public class ConsentList {
	public let entriesManager = EntriesManager()
	var publicKey: Data
	var privateKey: Data
	var identifier: String?
	var lastFetched: Date?
	var client: Client

	init(client: Client) {
		self.client = client
		privateKey = client.privateKeyBundleV1.identityKey.secp256K1.bytes
		publicKey = client.privateKeyBundleV1.identityKey.publicKey.secp256K1Uncompressed.bytes
		identifier = try? LibXMTP.generatePrivatePreferencesTopicIdentifier(privateKey: privateKey)
	}

	func load() async throws -> [ConsentListEntry] {
		guard let identifier = identifier else {
			throw ContactError.invalidIdentifier
		}
		let newDate = Date()

		let pagination = Pagination(
			limit: 500,
            after: lastFetched,
            direction: .ascending
        )
		let envelopes = try await client.apiClient.envelopes(topic: Topic.preferenceList(identifier).description, pagination: pagination)
    lastFetched = newDate

		var preferences: [PrivatePreferencesAction] = []

		for envelope in envelopes {
			let payload = try LibXMTP.userPreferencesDecrypt(publicKey: publicKey, privateKey: privateKey, message: envelope.message)

			try preferences.append(PrivatePreferencesAction(serializedData: Data(payload)))
		}
		for preference in preferences {
			for address in preference.allowAddress.walletAddresses {
				_ = await allow(address: address)
			}

			for address in preference.denyAddress.walletAddresses {
				_ = await deny(address: address)
			}

			for groupId in preference.allowGroup.groupIds {
				_ = await allowGroup(groupId: groupId)
			}

			for groupId in preference.denyGroup.groupIds {
				_ = await denyGroup(groupId: groupId)
			}
			
			for inboxId in preference.allowInboxID.inboxIds {
				_ = await allowInboxId(inboxId: inboxId)
			}

			for inboxId in preference.denyInboxID.inboxIds {
				_ = await denyInboxId(inboxId: inboxId)
			}
		}

		return await Array(entriesManager.map.values)
	}

    func publish(entries: [ConsentListEntry]) async throws {
      guard let identifier = identifier else {
        throw ContactError.invalidIdentifier
      }
      var payload = PrivatePreferencesAction()

      for entry in entries {
        switch entry.entryType {
		case .address:
		  switch entry.consentType {
		  case .allowed:
			payload.allowAddress.walletAddresses.append(entry.value)
		  case .denied:
			  payload.denyAddress.walletAddresses.append(entry.value)
		  case .unknown:
			  payload.messageType = nil
		  }
		case .group_id:
			switch entry.consentType {
			case .allowed:
				payload.allowGroup.groupIds.append(entry.value)
			case .denied:
				payload.denyGroup.groupIds.append(entry.value)
			case .unknown:
				payload.messageType = nil
    	    }
		case .inbox_id:
			switch entry.consentType {
			case .allowed:
			  payload.allowInboxID.inboxIds.append(entry.value)
			case .denied:
				payload.denyInboxID.inboxIds.append(entry.value)
			case .unknown:
				payload.messageType = nil
			}
		}
    }

    let message = try LibXMTP.userPreferencesEncrypt(
        publicKey: publicKey,
        privateKey: privateKey,
        message: payload.serializedData()
    )

    let envelope = Envelope(
        topic: Topic.preferenceList(identifier),
        timestamp: Date(),
        message: Data(message)
    )

    try await client.publish(envelopes: [envelope])
  }

	func allow(address: String) async -> ConsentListEntry {
		let entry = ConsentListEntry.address(address, type: ConsentState.allowed)
		await entriesManager.set(entry.key, entry)

		return entry
	}

	func deny(address: String) async -> ConsentListEntry {
		let entry = ConsentListEntry.address(address, type: ConsentState.denied)
		await entriesManager.set(entry.key, entry)

		return entry
	}

	func allowGroup(groupId: String) async -> ConsentListEntry {
		let entry = ConsentListEntry.groupId(groupId: groupId, type: ConsentState.allowed)
		await entriesManager.set(entry.key, entry)

		return entry
	}

	func denyGroup(groupId: String) async -> ConsentListEntry {
		let entry = ConsentListEntry.groupId(groupId: groupId, type: ConsentState.denied)
		await entriesManager.set(entry.key, entry)

		return entry
	}
	
	func allowInboxId(inboxId: String) async -> ConsentListEntry {
		let entry = ConsentListEntry.inboxId(inboxId, type: ConsentState.allowed)
		await entriesManager.set(entry.key, entry)

		return entry
	}

	func denyInboxId(inboxId: String) async -> ConsentListEntry {
		let entry = ConsentListEntry.inboxId(inboxId, type: ConsentState.denied)
		await entriesManager.set(entry.key, entry)

		return entry
	}

	func state(address: String) async -> ConsentState {
		guard let entry = await entriesManager.get(ConsentListEntry.address(address).key) else {
			return .unknown
		}

		return entry.consentType
	}

	func groupState(groupId: String) async -> ConsentState {
		guard let entry =  await entriesManager.get(ConsentListEntry.groupId(groupId: groupId).key) else {
			return .unknown
		}

		return entry.consentType
	}
	
	func inboxIdState(inboxId: String) async -> ConsentState {
		guard let entry = await entriesManager.get(ConsentListEntry.inboxId(inboxId).key) else {
			return .unknown
		}

		return entry.consentType
	}
}

/// Provides access to contact bundles.
public actor Contacts {
	var client: Client

	// Save all bundles here
	var knownBundles: [String: ContactBundle] = [:]

	// Whether or not we have sent invite/intro to this contact
	var hasIntroduced: [String: Bool] = [:]

	public var consentList: ConsentList

	init(client: Client) {
		self.client = client
		consentList = ConsentList(client: client)
	}

  public func refreshConsentList() async throws -> ConsentList {
		_ = try await consentList.load()
		return consentList
	}

	public func isAllowed(_ address: String) async -> Bool {
		return await consentList.state(address: address) == .allowed
	}

	public func isDenied(_ address: String) async -> Bool {
		return await consentList.state(address: address) == .denied
	}

	public func isGroupAllowed(groupId: String) async -> Bool {
		return await consentList.groupState(groupId: groupId) == .allowed
	}

	public func isGroupDenied(groupId: String) async -> Bool {
		return await consentList.groupState(groupId: groupId) == .denied
	}
	
	public func isInboxAllowed(inboxId: String) async -> Bool {
		return await consentList.inboxIdState(inboxId: inboxId) == .allowed
	}

	public func isInboxDenied(inboxId: String) async -> Bool {
		return await consentList.inboxIdState(inboxId: inboxId) == .denied
	}

	public func allow(addresses: [String]) async throws {
		var entries: [ConsentListEntry] = []

		try await withThrowingTaskGroup(of: ConsentListEntry.self) { group in
			for address in addresses {
				group.addTask {
					return await self.consentList.allow(address: address)
				}
			}

			for try await entry in group {
				entries.append(entry)
			}
		}
        try await consentList.publish(entries: entries)
	}

	public func deny(addresses: [String]) async throws {
		var entries: [ConsentListEntry] = []

		try await withThrowingTaskGroup(of: ConsentListEntry.self) { group in
			for address in addresses {
				group.addTask {
					return await self.consentList.deny(address: address)
				}
			}

			for try await entry in group {
				entries.append(entry)
			}
		}
        try await consentList.publish(entries: entries)
	}

	public func allowGroups(groupIds: [String]) async throws {
		var entries: [ConsentListEntry] = []

		try await withThrowingTaskGroup(of: ConsentListEntry.self) { group in
			for groupId in groupIds {
				group.addTask {
					return await self.consentList.allowGroup(groupId: groupId)
				}
			}

			for try await entry in group {
				entries.append(entry)
			}
		}
        try await consentList.publish(entries: entries)
	}

	public func denyGroups(groupIds: [String]) async throws {
		var entries: [ConsentListEntry] = []

		try await withThrowingTaskGroup(of: ConsentListEntry.self) { group in
			for groupId in groupIds {
				group.addTask {
					return await self.consentList.denyGroup(groupId: groupId)
				}
			}

			for try await entry in group {
				entries.append(entry)
			}
		}        
		try await consentList.publish(entries: entries)
	}
	
	public func allowInboxes(inboxIds: [String]) async throws {
		var entries: [ConsentListEntry] = []

		try await withThrowingTaskGroup(of: ConsentListEntry.self) { group in
			for inboxId in inboxIds {
				group.addTask {
					return await self.consentList.allowInboxId(inboxId: inboxId)
				}
			}

			for try await entry in group {
				entries.append(entry)
			}
		}
		try await consentList.publish(entries: entries)
	}

	public func denyInboxes(inboxIds: [String]) async throws {
		var entries: [ConsentListEntry] = []

		try await withThrowingTaskGroup(of: ConsentListEntry.self) { group in
			for inboxId in inboxIds {
				group.addTask {
					return await self.consentList.denyInboxId(inboxId: inboxId)
				}
			}

			for try await entry in group {
				entries.append(entry)
			}
		}
		try await consentList.publish(entries: entries)
	}

	func markIntroduced(_ peerAddress: String, _ isIntroduced: Bool) {
		hasIntroduced[peerAddress] = isIntroduced
	}

	func has(_ peerAddress: String) -> Bool {
		return knownBundles[peerAddress] != nil
	}

	func needsIntroduction(_ peerAddress: String) -> Bool {
		return hasIntroduced[peerAddress] != true
	}

	func find(_ peerAddress: String) async throws -> ContactBundle? {
		if let knownBundle = knownBundles[peerAddress] {
			return knownBundle
		}

		let response = try await client.query(topic: .contact(peerAddress))

		for envelope in response.envelopes {
			if let contactBundle = try? ContactBundle.from(envelope: envelope) {
				knownBundles[peerAddress] = contactBundle
				return contactBundle
			}
		}
		return nil
	}
}
